File: algoDat/public/code/vorlesung-01/cpp/Merge.cpp
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>

#include <vector>
#include "./Merge.h"

// ____________________________________________________________________________
std::vector<int> merge(std::vector<int> v1, std::vector<int> v2) {
  std::size_t n1 = v1.size();
  std::size_t n2 = v2.size();
  std::size_t i1 = 0;
  std::size_t i2 = 0;
  std::size_t k = 0;
  std::vector<int> result(n1 + n2);
  while (i1 < n1 || i2 < n2) {
    if (i1 < n1 && (i2 == n2 || v1[i1] <= v2[i2])) { result[k++] = v1[i1++]; }
    if (i2 < n2 && (i1 == n1 || v2[i2] <= v1[i1])) { result[k++] = v2[i2++]; }
  }
  return result;
}

File: algoDat/public/code/vorlesung-01/python/merge.py
"""
Copyright 2017, University of Freiburg.
Chair of Algorithms and Data Structures.
Hannah Bast <bast@cs.uni-freiburg.de>
"""

def merge(array1, array2):
    """
    
    Merge two already sorted arrays.

    >>> merge([2, 4, 6, 8], [1, 3, 5, 7])
    [1, 2, 3, 4, 5, 6, 7, 8]
    >>> merge([1, 2, 3, 4], [])
    [1, 2, 3, 4]
    >>> merge([], [1, 2, 3, 4])
    [1, 2, 3, 4]
    >>> merge([], [])
    []

    """

    n1 = len(array1)
    n2 = len(array2)
    i1 = 0
    i2 = 0
    result = []
    while i1 < n1 or i2 < n2:
        if i1 < n1 and (i2 == n2 or array1[i1] <= array2[i2]):
            result.append(array1[i1])
            i1 += 1
        if i2 < n2 and (i1 == n1 or array2[i2] <= array1[i1]):
            result.append(array2[i2])
            i2 += 1
    return result
